src/main/java/edu/harvard/iq/dataverse/LoginPage.java

```diff
@@ -129,16 +125,20 @@ public void init() {
         return infos;
     }
     
+    /**
+     * Retrieve information about all enabled identity providers in a sorted order to be displayed to the user.
+     * @return list of display information for each provider
+     */
     public List<AuthenticationProviderDisplayInfo> listAuthenticationProviders() {
         List<AuthenticationProviderDisplayInfo> infos = new LinkedList<>();
-        for (String id : authSvc.getAuthenticationProviderIdsSorted()) {
-            AuthenticationProvider authenticationProvider = authSvc.getAuthenticationProvider(id);
-            if (authenticationProvider != null) {
-                if (ShibAuthenticationProvider.PROVIDER_ID.equals(authenticationProvider.getId())) {
-                    infos.add(authenticationProvider.getInfo());
-                } else {
-                    infos.add(authenticationProvider.getInfo());
-                }
+        List<AuthenticationProvider> idps = new ArrayList<>(authSvc.getAuthenticationProviders());
+        
+        // sort by order first. in case of same order values, be deterministic in UI and sort by id, too.
+        Collections.sort(idps, Comparator.comparing(AuthenticationProvider::getOrder).thenComparing(AuthenticationProvider::getId));
+        
+        for (AuthenticationProvider idp : idps) {
+            if (idp != null) {
+                infos.add(idp.getInfo());
```

Comment:
Is the change above necessary for OIDC to work? I don't think so. It does look like a nice cleanup though. I like how it's deterministic. Perhaps this change should be documented.

Reply:
Actually yes it is. From this function, `authSvc.getAuthenticationProviderIdsSorted()` was called, which is deleted below. This was done in 3fd326c23 and here's what I wrote in the commit log:

> Refactor LoginPage.listAuthenticationProviders() to represent actual provider configuration.
>
> Before this refactoring, a static list of providers has been used, no matter if they where
enabled or not. This is not acceptable for the new OIDC provider, as we might have multiple
of 'em and it retrieves metadata on creation.
>
> So the idea was to not create some static list of IDs (which had been used to look up the real
provider from the "registry"), but instead get the registred providers, sort them and
retrieve the display info.
>
> Sorting is done by a numerical value first, then by id attribute (which is set by factories),
so we have a deterministic, unchanging order for the user.
>
> This lead to the creation of a new method AuthenticationProvider.getOrder(), defaulting to 1
and overridden by implementations. This opens up the possibility to change ordering by
configuration (as suggested in an old TODO comment of the refactored code)